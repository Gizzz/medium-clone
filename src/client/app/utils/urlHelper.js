import queryString from 'query-string';

const urlHelper = {
  computeEncodedRedirectUrl() {
    const decodedParams = queryString.parse(window.location.search);
    const hasRedirectUrl = Boolean(decodedParams.redirectUrl);

    let redirectUrl;
    if (hasRedirectUrl) {
      // eslint-disable-next-line prefer-destructuring
      redirectUrl = decodedParams.redirectUrl;
    } else {
      redirectUrl = window.location.pathname + window.location.search + window.location.hash;
    }

    const encodedRedirectUrl = window.encodeURIComponent(redirectUrl);
    return encodedRedirectUrl;
  },
};

export default urlHelper;
