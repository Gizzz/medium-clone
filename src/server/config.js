/**
 * Warning: in real-world app all sensitive data should be stored
 * in a safe place (.dotenv files for example)
 */

module.exports = {
  defaultPort: 3000,
  jwtSecret: 'secret',
};
