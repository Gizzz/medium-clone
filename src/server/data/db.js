const path = require('path');
const lowdb = require('lowdb');
const FileSync = require('lowdb/adapters/FileSync');

const adapter = new FileSync(path.resolve(__dirname, './db.json'));
const db = lowdb(adapter);

module.exports = db;
