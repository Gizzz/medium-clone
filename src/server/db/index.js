const path = require('path');
const lowdb = require('lowdb');
const FileSync = require('lowdb/adapters/FileSync');
const lodashId = require('lodash-id');

const dbFile = process.env.NODE_ENV === 'test'
  ? 'test-db-data.json'
  : 'db-data.json';

const adapter = new FileSync(path.resolve(__dirname, dbFile));
const db = lowdb(adapter);
db._.mixin(lodashId);

module.exports = db;
