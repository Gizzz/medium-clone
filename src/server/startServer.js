const path = require('path');
const express = require('express');
const compression = require('compression');
const bodyParser = require('body-parser');
const cors = require('cors');
const fs = require('fs');
const util = require('util');
//
const routes = require('./routes');
const config = require('./config');

/**
 * @description starts server and when started - returns server instance
 */
async function startServer(port) {
  const app = express();

  // middlewares

  app.use(cors());
  app.use(compression());
  app.use(bodyParser.json());

  app.use(express.static(
    path.resolve(__dirname, '../../dist'),
  ));

  // routes

  app.use('/api', routes);

  app.get('/*', (req, res) => {
    res.sendFile(path.resolve(__dirname, '../../dist/index.html'));
  });

  app.use((err, req, res, _next) => {
    console.error(`${err.stack}\n`);
    res.status(500).json({ error: 'Internal server error.' });
  });

  const mode = process.env.NODE_ENV ? process.env.NODE_ENV : 'development';
  const { defaultPort } = config;
  const actualPort = port || process.env.PORT || defaultPort;

  if (process.env.NODE_ENV === 'test') {
    console.log('Setting up test-db...');
    await createTestDbFile();
  }

  const server = await createServer(app, actualPort, mode);
  return server;
}

// helpers

async function createServer(app, port, mode) {
  return new Promise(resolve => {
    const server = app.listen(port, () => {
      console.log(`Medium-clone server is running on port ${port} (mode: ${mode})`);
      resolve(server);
    });
  });
}

async function createTestDbFile() {
  const readFile = util.promisify(fs.readFile);
  const writeFile = util.promisify(fs.writeFile);

  const originalData = await readFile(path.resolve(__dirname, 'db/db-data.json'), 'utf8');
  await writeFile(path.resolve(__dirname, 'db/test-db-data.json'), originalData);
}

module.exports = startServer;

/* eslint function-paren-newline: off */
