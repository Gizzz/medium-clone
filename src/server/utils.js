const _ = require('lodash');

function userToJson(user) {
  const safeUser = _.omit(user, ['password']);
  return safeUser;
}

module.exports = { userToJson };
