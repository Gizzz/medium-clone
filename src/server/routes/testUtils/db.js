import faker from 'faker';

import db from '../../db';
import dbData from '../../db/db-data.json';

const initDb = () => {
  db
    .setState(dbData)
    .write();
};

const generateRegistrationData = (overrides = {}) => {
  const password = faker.internet.password(6);

  return {
    username: faker.internet.userName(),
    password,
    confirmPassword: password,
    ...overrides,
  };
};

export { initDb, generateRegistrationData };

/* eslint import/no-extraneous-dependencies: off */
