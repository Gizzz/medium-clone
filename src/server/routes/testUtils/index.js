import faker from 'faker';
import bcrypt from 'bcrypt';

import db from '../../db';
import dbData from '../../db/db-data.json';

const initDb = () => {
  const dbDataCopy = JSON.parse(JSON.stringify(dbData));

  db
    .setState(dbDataCopy)
    .write();
};

const createUserInDb = async (userCredentials = {}, overrides = {}) => {
  // todo: this is copypasted from auth/createNewUser, think about removing duplication

  const username = userCredentials.username || faker.internet.userName();
  const password = userCredentials.password || faker.internet.password(6);

  const saltRounds = 1;
  const passwordHash = await bcrypt.hash(password, saltRounds);

  const maxUserId = db
    .get('users')
    .value()
    .map(user => user.id)
    .reduce((acc, cur) => (cur > acc ? cur : acc));

  const newUser = db
    .get('users')
    .insert({
      id: maxUserId + 1,
      username,
      passwordHash,
      avatarUrl: 'https://cdn-images-1.medium.com/fit/c/120/120/0*cmAOkoH29zoIVIBT',
      bio: '',
      ...overrides,
    })
    .write();

  return newUser;
};

const generateRegistrationData = (overrides = {}) => {
  const password = faker.internet.password(6);

  return {
    username: faker.internet.userName(),
    password,
    confirmPassword: password,
    ...overrides,
  };
};

export {
  initDb,
  createUserInDb,
  generateRegistrationData,
};

/* eslint import/no-extraneous-dependencies: off */
