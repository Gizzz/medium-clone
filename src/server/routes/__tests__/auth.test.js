import axios from 'axios';

import startServer from '../../startServer';
import config from '../../config';
import db from '../../db';
import { userToJson } from '../../utils';
import { initDb, createUserInDb, generateRegistrationData } from '../testUtils';

describe('auth', () => {
  const baseUrl = `http://localhost:${config.defaultPort}`;
  let server;

  beforeAll(async () => {
    server = await startServer();
  });

  afterAll(() => server.close());

  beforeEach(() => {
    initDb();
  });

  test('login - should require username and password', async () => {
    const err1 = await axios
      .post(`${baseUrl}/api/auth/login`, {})
      .catch(e => e);

    expect(err1.response.status).toBe(400);
    expect(err1.response.data.error).toMatchSnapshot();

    const err2 = await axios
      .post(`${baseUrl}/api/auth/login`, { password: '123123' })
      .catch(e => e);

    expect(err2.response.status).toBe(400);
    expect(err2.response.data.error).toMatchSnapshot();

    const err3 = await axios
      .post(`${baseUrl}/api/auth/login`, { username: 'test' })
      .catch(e => e);

    expect(err3.response.status).toBe(400);
    expect(err3.response.data.error).toMatchSnapshot();
  });

  test('login - should return 404 if username does not exist', async () => {
    const err = await axios
      .post(`${baseUrl}/api/auth/login`, {
        username: '__this_name_should_not_exist__',
        password: '123123',
      })
      .catch(e => e);

    expect(err.response.status).toBe(404);
    expect(err.response.data.error).toMatchSnapshot();
  });

  test('login - should check password hash', async () => {
    // create test user to check password against
    //
    const registrationData = generateRegistrationData();
    await createUserInDb({
      username: registrationData.username,
      password: registrationData.password,
    });

    const wrongPassword = registrationData.password.repeat(2);
    const err = await axios
      .post(`${baseUrl}/api/auth/login`, {
        username: registrationData.username,
        password: wrongPassword,
      })
      .catch(e => e);

    expect(err.response.status).toBe(400);
    expect(err.response.data.error).toMatchSnapshot();
  });

  test('login - should return user with token on success', async () => {
    const registrationData = generateRegistrationData();
    const userFromDb = await createUserInDb({
      username: registrationData.username,
      password: registrationData.password,
    });

    const loginResponse = await axios
      .post(`${baseUrl}/api/auth/login`, {
        username: registrationData.username,
        password: registrationData.password,
      })
      .then(res => res.data);

    expect(loginResponse.user).toEqual(userToJson(userFromDb));
    expect(loginResponse.user.passwordHash).not.toBeDefined();
    expect(loginResponse.token).toBeDefined();
    expect(typeof loginResponse.token).toBe('string');
  });

  test('register - should fail on request with empty payload', async () => {
    // no payload object
    const err1 = await axios
      .post(`${baseUrl}/api/auth/register`)
      .catch(e => e);

    expect(err1.response.status).toBe(400);
    expect(err1.response.data.error).toMatchSnapshot();

    // empty payload object
    const err2 = await axios
      .post(`${baseUrl}/api/auth/register`, {})
      .catch(e => e);

    expect(err2.response.status).toBe(400);
    expect(err2.response.data.error).toMatchSnapshot();
  });

  test('register - should require all fields', async () => {
    const err1 = await axios
      .post(`${baseUrl}/api/auth/register`, { username: 'test' })
      .catch(e => e);

    expect(err1.response.status).toBe(400);
    expect(err1.response.data.error).toMatchSnapshot();

    const err2 = await axios
      .post(`${baseUrl}/api/auth/register`, { username: 'test', password: '123123' })
      .catch(e => e);

    expect(err2.response.status).toBe(400);
    expect(err2.response.data.error).toMatchSnapshot();

    const err3 = await axios
      .post(`${baseUrl}/api/auth/register`, { username: 'test', confirmPassword: '123123' })
      .catch(e => e);

    expect(err3.response.status).toBe(400);
    expect(err3.response.data.error).toMatchSnapshot();
  });

  test('register - should reject if username is too short', async () => {
    const err = await axios
      .post(`${baseUrl}/api/auth/register`, { username: 'ts', password: '123123', confirmPassword: '123123' })
      .catch(e => e);

    expect(err.response.status).toBe(400);
    expect(err.response.data.error).toMatchSnapshot();
  });

  test('register - should reject if password is too short', async () => {
    const err = await axios
      .post(`${baseUrl}/api/auth/register`, { username: 'test', password: '12345', confirmPassword: '12345' })
      .catch(e => e);

    expect(err.response.status).toBe(400);
    expect(err.response.data.error).toMatchSnapshot();
  });

  test('register - should reject if password and confirmation do not match', async () => {
    const err = await axios
      .post(`${baseUrl}/api/auth/register`, { username: 'test', password: '123123', confirmPassword: '456456' })
      .catch(e => e);

    expect(err.response.status).toBe(400);
    expect(err.response.data.error).toMatchSnapshot();
  });

  test('register - should reject if username is already taken', async () => {
    // create test user to check username against
    //
    const registrationData = generateRegistrationData();
    await createUserInDb({
      username: registrationData.username,
      password: registrationData.password,
    });

    const err = await axios
      .post(`${baseUrl}/api/auth/register`, registrationData)
      .catch(e => e);

    expect(err.response.status).toBe(400);
    expect(err.response.data.error).toMatchSnapshot();
  });

  test('register - should create and return user with token on success', async () => {
    const registrationData = generateRegistrationData();
    const registrationResponse = await axios
      .post(`${baseUrl}/api/auth/register`, registrationData)
      .then(res => res.data);

    const userFromDb = db
      .get('users')
      .getById(registrationResponse.user.id)
      .value();

    expect(registrationResponse.user).toEqual(userToJson(userFromDb));
    expect(registrationResponse.user.passwordHash).not.toBeDefined();
    expect(registrationResponse.token).toBeDefined();
    expect(typeof registrationResponse.token).toBe('string');
  });
});
