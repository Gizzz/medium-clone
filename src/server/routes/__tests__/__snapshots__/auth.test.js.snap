// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth login - should check password hash 1`] = `"Password is wrong."`;

exports[`auth login - should require username and password 1`] = `"Username and password should be provided."`;

exports[`auth login - should require username and password 2`] = `"Username and password should be provided."`;

exports[`auth login - should require username and password 3`] = `"Username and password should be provided."`;

exports[`auth login - should return 404 if username does not exist 1`] = `"User not found. Check the username."`;

exports[`auth register - should fail on request with empty payload 1`] = `"All fields are required."`;

exports[`auth register - should fail on request with empty payload 2`] = `"All fields are required."`;

exports[`auth register - should reject if password and confirmation do not match 1`] = `"Password and confirmation do not match."`;

exports[`auth register - should reject if password is too short 1`] = `"Password should be at least 6 symbols long."`;

exports[`auth register - should reject if username is already taken 1`] = `"This username is already taken."`;

exports[`auth register - should reject if username is too short 1`] = `"Username should be at least 3 symbols long."`;

exports[`auth register - should require all fields 1`] = `"All fields are required."`;

exports[`auth register - should require all fields 2`] = `"All fields are required."`;

exports[`auth register - should require all fields 3`] = `"All fields are required."`;
